lambda/index.py

```diff
@@ -184,40 +234,33 @@ def main(event, context):
-                next_log_group, ok = modify_subscriptions(
-                    logsClient, False, matches, exclusions, start_log_group, args)
-
-            if ok:
-                if next_log_group is None:
-                    cfnresponse.send(cfnEvent, context,
-                                     cfnresponse.SUCCESS, {})
-                else:
-                    logging.info(
-                        'sending pagination event: next_log_group=%s', next_log_group)
-                    entry = {
-                        'Time': datetime.datetime.now(),
-                        'Source': EVENTBRIDGE_SOURCE,
-                        'DetailType': EVENTBRIDGE_DETAIL_TYPE,
-                        'Detail': json.dumps({
-                            'cfnEvent': cfnEvent,
-                            'next': next_log_group,
-                        }),
-                    }
-                    eventsClient.put_events(Entries=[entry])
-            else:
-                data = {
-                    'Data': 'Error: unable to create subscriptions for any log groups',
-                }
-                cfnresponse.send(cfnEvent, context, cfnresponse.FAILED, data)
+            # This code exists so that lambda failures don't fail silently and indefinitely block
+            # the CloudFormation stack creation progress. Instead, this code tries to make it so that users
+            # actually get feedback if something goes wrong, saving them ~30 minutes.
+            #
+            # If the main thread completes in time, the lambda exits once the main thread is done
+            # and a successful response is sent to CloudFormation.
+            # If the main thread doesn't complete in time, cancel_thread hopefully completes
+            # and sends a response to CloudFormation before the Lambda execution environment is killed.
+            #
+            # If the cancel thread completes, then it's likely that the CloudFormation stack that
+            # calls this code will trigger a rollback. Deletion is likely to time out as well since
+            # the same code path is executed, resulting in an incomplete cleanup. Incomplete cleanup
+            # is fine, since the lambda code knows how to deal with it. A user just needs to rerun the
+            # CloudFormation stack or Terraform module with a larger timeout.
+            #
+            # There is a chance of a race condition where we send 2 responses to CloudFormation. This
+            # case is unlikely. It also doesn't result in incomplete setup for the customer.
+            main_thread = threading.Thread(target=process_setup_event, args=(
+                logsClient, eventsClient, cfnEvent, start_log_group, matches, exclusions, args, context))
+            cancel_thread = threading.Thread(
+                target=send_cfnresponse_5s_before_timeout, args=(timeout, cfnEvent, context))
+            main_thread.start()
+            cancel_thread.start()
+            main_thread.join()
```

Comment:
Can we call join() with a timeout and use is_alive() after to determine if main_thread timed out? I don't know if the threading library is bad, but it seems like that's what we're doing here

Reply:
I think the configuration you have here that bails 5 seconds early to do actual cleanup makes sense, it just seems like you could call join(timeout - 5) and still have that same 5 seconds to do the same cleanup. I agree we shouldn't wait until AWS takes action against us. Do you have an environment to test the timeout?

Reply:
if we call join() with timeout, then we hurt the performance in the normal case. The function gets executed sequentially a few times in that case, meaning that the normal latency would now be ~(120 - 5) * 5 seconds or so.

The thing I used to test was this, with timeout altered a couple of times:
# resource "aws_cloudformation_stack" "this" {
#   name = "luke-autosubscribe-test2"

#   template_body = file("/home/luke/github.com/terraform-aws-cloudwatch-logs-subscription/cloudformation/generated/subscribelogs.yaml")

#   capabilities = ["CAPABILITY_IAM"]

#   parameters = {
#     "CollectionStackName" = "luke-autosubscribe-test1"
#     "LogGroupMatches" = ".*"
#     "FilterName" = "luke-test-filter-pattern"
#     "LogGroupExpirationInDays" = 7
#     "LambdaTimeout" = 5
#     "DestinationArnOverride" = module.cloudwatch_kinesis_firehose.firehose_delivery_stream.arn
#     "DeliveryStreamRoleArnOverride" = aws_iam_role.subscription_filter.arn
#   }
# }

Reply:
OK that makes sense, python has some sort of retry logic which isn't what we want. Looks good to me